### ISSUE TITLE
TCP request and tcpin nodes still facing reconnect issue

### ISSUE THREAD

**Issue by taralshah123** [node-red/node-red#1469]
First of all, thanks for this awesome nodered tool, which enables us non JavaScript coders to integrate this technology in our embedded systems environment. I am using this from 2 years. I am facing issue in TCP in and tcp request node for reconnect. I already saw the issues #759 #150. Sorry to say, but these issues were in 2014 around so I think it's fix should be already committed in latest version. Or if manual needs to be fixed, i do not know java script much. 
My issue is it connects successful to my server (esp8266). Talks perfectly. But if esp8266 goes down and come back then my nodered doesn't detect this event. It feels it's already connected and doesn't reestablish connection. 
But on the other end if I started nodered and that time if I keep esp8266 down then it detects no connection and try to connect. As soon as esp8266 is up, it connects.
So I think issue is not of connecting or reconnecting. The issue is detection of disconnected event. 
My application scenario: nodered connection to esp8266 (running as tcp server).
Can anyone please help me to solve it if already faced?
Sorry for if any technical mistakes, I am not from javascript background. I searched so many forums and still not getting solution so submitted issue.

**Comment by hardillb** [comment]
How long have you left this after restarting the esp? The normal TCP
timeout is 15mins.

So if the esp is just sending data, the node may end up waiting the full
15mins before noticing the connection has dropped.

**Comment by taralshah123** [comment]
Thanks for your quick response.
I was restarting esp in 3-4 minutes. So as you say if this time takes 15 minutes, nodered will detect disconnect right? So by changing timeout configuration in settings.js will it be solved? I think I tried by changing then to 1 minute or so. Still i will test again that tonight and post back with results.

**Comment by taralshah123** [comment]
Hello hardillb...
I tried to experiment the same. I connect esp to nodered. It connects. Then after I turn off esp, i waited till 20 minutes, but nodered still says it's connected. Is there anyone facing same or can anyone test this scenario with their environment?

**Comment by taralshah123** [comment]
Even now after 55 minutes it still shows green dot, status connected.

**Comment by beneuto** [comment]
I have the same issue.   Randomly, data will stop flowing from TCP IN node.   Redeploying flow fixes it.

**Comment by DenDeze** [comment]
Same issue here... also esp as tcp server. restart wifi, no data received after 1 hour. redeploy flow fixes the issue. latest node red version.

**Comment by HarmOtten** [comment]
Same issue here as well. After rebooting my router, the tcp-in-node still thinks it is connected, but data is not coming in. Even after a few hours.
Apart from this, the tcp-out-node seems to "fall asleep" after a few hours. The first one or two messages send after a few hours often does not reach the destination. (I use the "Connect to port ... at host ..." type, and "Close connection after each message is sent?" is checked.)
I hope these issues can be solved, because I love working with node-RED!

**Comment by jeveloper** [comment]
@HarmOtten  Totally off topic, i thought that 1.0 would be out by now, it was announced on a roadmap back in July , 2017, do you think IBM has abandoned this project? Thanks , trying to evaluate if its worth investing time.

**Comment by knolleary** [comment]
@jeveloper Hi there - yes that is totally off topic. You could ask on the very active [mailing list](https://groups.google.com/forum/#!forum/node-red) or [slack team](https://nodered.org/slack). (Admittedly, it has been slightly quieter over the Christmas period).
You could also see the commit activity on the master, 0.18 and projects branches to see there's lots going on. Yes 1.0 is behind schedule - I'll be doing an update blog post in the near future.

**Comment by jeveloper** [comment]
@knolleary Thanks for responding and so quickly ! Appreciate that.
I figured i'd ask someone who has production experience and problem hence partial and with experience.  Thanks

**Comment by Mitha17** [comment]
Hi, am trying to send my data through esp (server) to my node red.. tcp is getting connected and wifi send my message to.. but the data is not shown in my node red..can anyone help me please?

**Comment by HarmOtten** [comment]
@Mitha17 , How are you sending the data from your ESP? What node are you using to receive the data in Node-RED? Is this related to the reconnect problems in this github-issue? If not, I think the best place for help is: [https://groups.google.com/forum/#!forum/node-red]

**Comment by HarmOtten** [comment]
I managed to work around the tcp-out-node reconnect issue, by sending a message at least once every hour. This seems to keep the connection open. But for the tcp-in node I cannot do that. When is has lost the connection, the only thing I can do to reconnect it is to change the node (for instance, just move it around) and then deploy it again.
Is there any progress in solving this issue?

**Comment by dceejay** [comment]
Hi - must admit been rather busy on the 0.18 release and associated bits and pieces...
Are you using it as a server or client ?

**Comment by HarmOtten** [comment]
Well, I have a ESP8266 running esp-link-v2.2.3 that (tries to) send serial messages to a tcp-in-node in Node-RED. Does that answer your question?

**Comment by dceejay** [comment]
ah so you have Node-RED set to listen On... 
so yes a server - so ignore previous suggestion :-)

**Comment by dceejay** [comment]
actually can you share both your in and out nodes here please ? Export just those nodes. just so I've can see config of both in and out. Thanks

**Comment by HarmOtten** [comment]
Sure!
Here are my tcp-in and out nodes:
`[{"id":"3e66be14.d63172","type":"tcp in","z":"9caf42.66b420c","name":"OTGW","server":"client","host":"192.168.1.12","port":"23","datamode":"stream","datatype":"utf8","newline":"\\n","topic":"","base64":false,"x":70,"y":80,"wires":[["150cee4e.f575c2","1daa900c.9b8fd"]]},{"id":"5eaa62e9.d92acc","type":"tcp out","z":"9caf42.66b420c","host":"192.168.1.12","port":"23","beserver":"client","base64":false,"end":true,"name":"OTGW","x":850,"y":460,"wires":[]}]`

**Comment by dceejay** [comment]
Ah - no you are connecting to,,,, (client mode)... so as you were... please try that patch suggested above !...
Thanks

**Comment by HarmOtten** [comment]
Do you mean, I should upgrade my Node-RED to the latest version? (Currently running 0.17.5)

**Comment by dceejay** [comment]
No, 2 hours ago I suggested editing the TCP node file. See above

**Comment by HarmOtten** [comment]
Sorry, I don't understand what you mean. When I set the tcp-in-node to "Listen on", I cannot configure the ip-address that it needs to connect to, and results in error: "unable to listen on port 23, error: Error: listen EACCES 0.0.0.0:23".
Or do you mean something else? Please explain clearly. Thank you!

**Comment by mscbpi** [comment]
@HarmOtten he suggests editing 31-tcpin.js that resides in your nodered installation in node_modules folder.

I am facing the exact same issue for connection to a moxa rs232 gateway that is not reachable permanently  from  the nodered instance (as designed, moving part out of network scope).

 I may try the fix when I get a permanent access to the Moxa. Meanwhile I have configured the moxa as UDP streamer rather than TCP server.

**Comment by dceejay** [comment]
@harmotten - the nodes you sent me are set 
![image](https://user-images.githubusercontent.com/5375409/35849182-a5094ba8-0b18-11e8-845d-002996e9eb91.png)
so they are "connecting to" your other devices so are in client mode.
The patch I'd like you to test is
find the file 31-tcpin.js
insert at line 55

    client.setKeepAlive(true,120000);

and likewise the same at line 243

**Comment by HarmOtten** [comment]
@dceejay, I missed your edited comments. Somehow they were not updated on my screen. Now it makes sense ;-).
I found the file at: /usr/lib/node_modules/node-red/nodes/core/io/31-tcpin.js
At line 55 I see: `connectionPool[id] = client;`
Do you mean inserting before or after this line?
Same for line 243: `client.on('end', function (err) {`

**Comment by knolleary** [comment]
Closing as part of a tidy-up of the issue tracker — this has been inactive for a long time and much has changed in Node-RED since. If it's still relevant on a current version, please open a fresh issue with up-to-date details and we'll take a look. Thanks for raising it.
